[MISC] Clarify participant_id in participants.tsv file if it exists

adam2392 opened the pull request:
Closes: #737 

Question: "In case of single-session studies, this file has one compulsory column participant_id that consists of sub-<label>, followed by a list of optional columns describing participants. "  is in the current spec.

What does that mean? In case of single-session studies? Does that mean in single-session studies, this file is compulsory with participant_id a certain way?

sappelhoff commented:
Thanks for taking a stab at this @adam2392 --> I think it's good to clarify it.

With regards to your question --> I have no clue. Perhaps the other @bids-standard/maintainers know.

I had a look at the [Longitudinal and multi-site studies](https://bids-specification.readthedocs.io/en/latest/06-longitudinal-and-multi-site-studies.html) section to see whether it clarifies things, but I think it should rather be clarified itself (using SHOULD / MUST terminology).

Let me give a few examples of what I think should change:

- Session label ~can~ MUST consist only of alphanumeric characters [a-zA-Z0-9] and ~should~ SHOULD be consistent across subjects.
- If numbers are used in session labels we ~recommend~ RECOMMEND using zero padding 
- Acquisition time of session can be defined in the sessions file ~(see below for details)~ [**insert markdown link for "sessions file"**]
- The extra session layer (at least one /ses-<label> subfolder) ~should~ MUST be added for all subjects if at least one subject in the dataset has more than one session. ~Skipping the session layer for only some subjects in the dataset is not allowed.~


And then under [Sessions file](https://bids-specification.readthedocs.io/en/latest/06-longitudinal-and-multi-site-studies.html#sessions-file) I find it confusing that the text refers to a "participant key file" ... probably, `sessions.tsv` is meant by that, so that should be clarified as well IMHO

> In case of multiple sessions there is an option of adding additional participant key files describing variables changing between sessions. In such case one file per participant should be added. These files need to include compulsory session_id column and describe each session by one and only one row. Column names in per participant key files have to be different from group level participant key column names.

adam2392 commented:
@sappelhoff 
- addressed all your comments, and 
- assuming the participants.tsv file is always RECOMMENDED, I added explicit language to clarify the question I had.

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -1,18 +1,17 @@
+The extra session layer (at least one `/ses-<label>` subfolder) SHOULD
 be added for all subjects if at least one subject in the dataset has more than
-one session. Skipping the session layer for only some subjects in the dataset is
-not allowed. If a `/ses-<label>` subfolder is included as part of the
-directory hierarchy, then the same `ses-<label>` tag must also be
+one session. If a `/ses-<label>` subfolder is included as part of the
+directory hierarchy, then the same `ses-<label>` tag MUST also be
```
> Skipping the session layer for only some subjects in the dataset is
not allowed

I actually think that the original text here meant to say that:

> The extra session layer (at least one `/ses-<label>` subfolder) MUST
be added for all subjects if at least one subject in the dataset has more than
one session.

instead of the SHOULD you used here.

opinions @Remi-Gau @tsalo ?

sappelhoff replied:
on the other hand, that doesn't really make sense, because there are experiments where some subjects perform an extra session :thinking:

adam2392 replied:
I think skipping the `ses-<X>` in some subjects is allowed if say not all subjects have the same sessions. Rn this results in a warning in validator, which basically all my data has this warning :p.

```The extra session layer (at least one /ses-<label> subfolder) MUST
be added for all subjects if at least one subject in the dataset has more than
one session.```

seems reasonable to me... The problem is how does one validate this? If one has multiple "sessions" per subject, then that implies there are potentially different electrode/channels across those sessions, so the user would want to store it as such. Does this basically mean if one has:

```
sub-01/
   sub-01_ses-01_task-blah_ieeg.vhdr
   ...
```
it's allowed, but if you have:


```
sub-01/
   sub-01_ses-01_task-blah_ieeg.vhdr
   sub-01_ses-02_task-blah_ieeg.vhdr
   ...
```
it's not?

sappelhoff replied:
> I think skipping the ses-<X> in some subjects is allowed if say not all subjects have the same sessions. Rn this results in a warning in validator, which basically all my data has this warning

yes, I experienced that warning multiple times myself :roll_eyes: I think this case is more common than thought when this warning was originally implemented. That is, more often it's a choice, rather than an overlooked error.

re: your example --> yes :thinking: but I think your current changes (apart from my comments below) are actually fine and clarify the situation without changing the state of anything (which is what we should aim for), so I'd be fine with leaving the SHOULD, not interpreting it as a MUST post-hoc.

Remi-Gau replied:
Wait wasn't the original intent of theb MUST to avoid things like:

```
├── sub-1
│   └── eeg
└── sub-2
    ├── ses-1
    │   └── eeg
    └── ses-2
        └── fmri
```

Or maybe I am missing something?

sappelhoff replied:
yes, that's how I interpret it as well. What's annoying is that first the text says "should" (not MUST), and then the follow up sentence is:

> Skipping the session layer for only some subjects in the dataset is not allowed.

so basically the "should" is actually a MUST.

However:

1. I am not sure why this rule makes sense
2. I am pretty sure that I have seen datasets not conforming to this rule
3. There is no validator support for this rule as far as I know

:roll_eyes: perhaps we can discuss this tonight in the maintainers meeting

Remi-Gau replied:
Yeah the inconsistency should be fixed or clarified.

I mean I find the folder structure I described above "ugly", but thank goodness we don't just rely on my sense of aesthetics.

:wink:

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
-participant key files describing variables changing between sessions. In such
-case one file per participant should be added. These files need to include
+``sessions.tsv`` files describing variables changing between sessions. In such
+case one file per participant should be added. These files MUST include
```
```suggestion
case one file per participant SHOULD be added. These files MUST include a
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
-case one file per participant should be added. These files need to include
+``sessions.tsv`` files describing variables changing between sessions. In such
+case one file per participant should be added. These files MUST include
 compulsory `session_id` column and describe each session by one and only one
```
```suggestion
`session_id` column and describe each session by one and only one row.
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
 compulsory `session_id` column and describe each session by one and only one
-row. Column names in per participant key files have to be different from group
-level participant key column names.
+row. Column names in per ``sessions.tsv`` files MUST be different from group
+level participant key column names in the ``participants.tsv`` file.
```
```suggestion
Column names in `sessions.tsv` files MUST be different from group
level participant key column names in the `participants.tsv` file.
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
 In case of multiple sessions there is an option of adding additional
-participant key files describing variables changing between sessions. In such
-case one file per participant should be added. These files need to include
+``sessions.tsv`` files describing variables changing between sessions. In such
```
```suggestion
`sessions.tsv` files describing variables changing between sessions. In such
```

sappelhoff commented on src/03-modality-agnostic-files.md:
```diff
@@ -162,9 +162,9 @@ participants.json
-In case of single-session studies, this file has one compulsory column
-`participant_id` that consists of `sub-<label>`, followed by a list of optional
-columns describing participants.
+In case of single-session studies, if this file exists, this file MUST contain the column
```
```suggestion
If this file exists, it file MUST contain the column
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -1,18 +1,17 @@
-and file names in the form of `ses-<label>`. Session label can consist
-only of alphanumeric characters `[a-zA-Z0-9]` and should be consistent across
-subjects. If numbers are used in session labels we recommend using zero padding
+and file names in the form of `ses-<label>`. Session label MUST consist
```
```suggestion
and file names in the form of `ses-<label>`. Session labels MUST consist
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
-level participant key column names.
+`sessions.tsv` files describing variables changing between sessions. In such
+case one file per participant SHOULD be added. These files MUST include a
+`session_id` column and describe each session by one and only one row.
+Column names in `sessions.tsv` files MUST be different from group
+level participant key column names in the `participants.tsv` file.
```
following from "participants.tsv":

> that MUST consist of `sub-<label>` identifying each row for a
participant, followed by a list of optional columns describing participants.

```suggestion
`session_id` column that MUST consist of `ses-<label>`, and that
MUST describe each session by one and only one row.
Column names in `sessions.tsv` files MUST be different from group
level participant key column names in the `participants.tsv` file.
```

sappelhoff replied:
could you please also add a link back to "participants file" in the last two words of this sentence?

adam2392 replied:
Done

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,11 @@ sub-<label>/
+case one file per participant SHOULD be added. These files MUST include a
+`session_id` column and describe each session by one and only one row.
+Column names in `sessions.tsv` files MUST be different from group
+level participant key column names in the [`participants.tsv` file](./03-modality-agnostic-files.md).
```
```suggestion
level participant key column names in the
[`participants.tsv` file](./03-modality-agnostic-files.md#participants-file).
```

tsalo commented on src/03-modality-agnostic-files.md:
```diff
@@ -162,9 +162,9 @@ participants.json
-`participant_id` that consists of `sub-<label>`, followed by a list of optional
-columns describing participants.
+If this file exists, it file MUST contain the column
+`participant_id` that MUST consist of `sub-<label>` identifying each row for a
```
```suggestion
`participant_id`, which MUST consist of `sub-<label>` values identifying one row for each
```

tsalo commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -1,18 +1,17 @@
-not allowed. If a `/ses-<label>` subfolder is included as part of the
-directory hierarchy, then the same `ses-<label>` tag must also be
+one session. If a `/ses-<label>` subfolder is included as part of the
+directory hierarchy, then the same `ses-<label>` tag MUST also be
```
```suggestion
directory hierarchy, then the same [`ses-<label>`](../99-appendices/09-entities.md#ses) key/value pair MUST also be
```

tsalo reviewed (changes requested):
Some minor wording/formatting request.

tsalo reviewed (approved):
The changes look good to me. Thanks!

effigies reviewed (approved):
LGTM. Only remaining suggestion is to split sentences with newlines.

sappelhoff commented on src/03-modality-agnostic-files.md:
```diff
@@ -162,9 +162,9 @@ participants.json
-In case of single-session studies, this file has one compulsory column
-`participant_id` that consists of `sub-<label>`, followed by a list of optional
-columns describing participants.
+If this file exists, it MUST contain the column
+`participant_id`, which MUST consist of `sub-<label>` values identifying one row for each
+participant, followed by a list of optional columns describing participants.
```
```suggestion
If this file exists, it MUST contain the column `participant_id`,
which MUST consist of `sub-<label>` values identifying one row for each participant,
followed by a list of optional columns describing participants.
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -1,18 +1,17 @@
-and file names in the form of `ses-<label>`. Session label can consist
-only of alphanumeric characters `[a-zA-Z0-9]` and should be consistent across
-subjects. If numbers are used in session labels we recommend using zero padding
+and file names in the form of `ses-<label>`. Session labels MUST consist
+only of alphanumeric characters `[a-zA-Z0-9]` and SHOULD be consistent across
+subjects. If numbers are used in session labels we RECOMMEND using zero padding
```
```suggestion
and file names in the form of `ses-<label>`.
Session labels MUST consist only of alphanumeric characters `[a-zA-Z0-9]`
and SHOULD be consistent across subjects.
If numbers are used in session labels we RECOMMEND using zero padding
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -1,18 +1,17 @@
-one session. Skipping the session layer for only some subjects in the dataset is
-not allowed. If a `/ses-<label>` subfolder is included as part of the
-directory hierarchy, then the same `ses-<label>` tag must also be
+one session. If a `/ses-<label>` subfolder is included as part of the
+directory hierarchy, then the same [`ses-<label>`](./99-appendices/09-entities.md#ses) key/value pair MUST also be
```
```suggestion
one session.
If a `/ses-<label>` subfolder is included as part of the directory hierarchy,
then the same [`ses-<label>`](./99-appendices/09-entities.md#ses)
key/value pair MUST also be
```

sappelhoff commented on src/06-longitudinal-and-multi-site-studies.md:
```diff
@@ -71,11 +70,12 @@ sub-<label>/
-compulsory `session_id` column and describe each session by one and only one
-row. Column names in per participant key files have to be different from group
-level participant key column names.
+`sessions.tsv` files describing variables changing between sessions. In such
+case one file per participant SHOULD be added. These files MUST include a
+`session_id` column and describe each session by one and only one row.
+Column names in `sessions.tsv` files MUST be different from group
+level participant key column names in the
+[`participants.tsv` file](./03-modality-agnostic-files.md#participants-file).
```
```suggestion
`sessions.tsv` files describing variables changing between sessions.
In such case one file per participant SHOULD be added.
These files MUST include a `session_id` column and describe each session by one and only one row.
Column names in `sessions.tsv` files MUST be different from group level participant key column names in the
[`participants.tsv` file](./03-modality-agnostic-files.md#participants-file).
```

sappelhoff commented:
Thanks @adam2392!
